homeassistant/components/google_generative_ai_conversation/conversation.py

```diff
@@ -140,133 +140,150 @@ async def async_process(
+                    f"Sorry, I had a problem with my template: {err}",
                 )
                 return conversation.ConversationResult(
                     response=intent_response, conversation_id=conversation_id
                 )
+            trace_element = trace.TraceElement({"prompt": prompt}, "google/prompt")
+            trace.trace_append_element(trace_element)
+
+            messages[0] = {"role": "user", "parts": prompt}
+            messages[1] = {"role": "model", "parts": "Ok"}
+
+            LOGGER.debug("Input: '%s' with history: %s", user_input.text, messages)
+
+            chat = model.start_chat(history=messages)
+            chat_request = user_input.text
+            # To prevent infinite loops, we limit the number of iterations
+            for _iteration in range(MAX_TOOL_ITERATIONS):
+                try:
+                    chat_response = await chat.send_message_async(chat_request)
+                except (
+                    ClientError,
+                    ValueError,
+                    genai_types.BlockedPromptException,
+                    genai_types.StopCandidateException,
+                ) as err:
+                    LOGGER.error("Error sending message: %s", err)
+                    intent_response.async_set_error(
+                        intent.IntentResponseErrorCode.UNKNOWN,
+                        f"Sorry, I had a problem talking to Google Generative AI: {err}",
+                    )
+                    return conversation.ConversationResult(
+                        response=intent_response, conversation_id=conversation_id
+                    )
 
-            LOGGER.debug("Response: %s", chat_response.parts)
-            if not chat_response.parts:
-                intent_response.async_set_error(
-                    intent.IntentResponseErrorCode.UNKNOWN,
-                    "Sorry, I had a problem talking to Google Generative AI. Likely blocked",
+                LOGGER.debug("Response: %s", chat_response.parts)
+                if not chat_response.parts:
+                    intent_response.async_set_error(
+                        intent.IntentResponseErrorCode.UNKNOWN,
+                        "Sorry, I had a problem talking to Google Generative AI. Likely blocked",
+                    )
+                    return conversation.ConversationResult(
+                        response=intent_response, conversation_id=conversation_id
+                    )
+                self.history[conversation_id] = chat.history
+                tool_call = chat_response.parts[0].function_call
+
+                if not tool_call or not llm_api:
+                    break
+                trace_element = trace.TraceElement(
+                    {
+                        "tool_name": tool_call.name,
+                        "tool_args": dict(tool_call.args),
+                    },
+                    "google/tool_call",
```

Comment:
I wonder if we should aim for generic conversation tracing or for integration specific. And for example, OpenAI can include multiple tool calls.